packages/snjs/lib/services/api/mfa_service.ts

```diff
@@ -0,0 +1,126 @@
+import { Challenge, ChallengeKeyboardType, ChallengePrompt, ChallengeReason, ChallengeValidation } from '@Lib/challenges';
+import { AuthMethods } from '@standardnotes/auth';
+import { ChallengeService } from '..';
+import { PureService } from '../pure_service';
+import { PromptTitles, SessionStrings } from './messages';
+import { HttpResponse } from './responses';
+
+export enum MfaStatus {
+  EnteredCode = 'mfa-entered-code',
+  NotRequired = 'mfa-not-required',
+
+  ErrorMissingCode = 'mfa-error-missing-code',
+  ErrorResponse = 'mfa-error-response',
+}
+
+export type MfaResult = MfaResultSuccess | MfaResultError
+
+export type MfaResultSuccess = {
+  status: MfaStatus.EnteredCode,
+  payload: MfaPayload,
+} | {
+  status: MfaStatus.NotRequired,
+}
+
+export type MfaResultError = {
+  status: MfaStatus.ErrorResponse,
+  response: HttpResponse,
+} | {
+  status: MfaStatus.ErrorMissingCode,
+}
+
+export type MfaPayload = {
+  mfaSecretUuid: string,
+  enteredTotp: string,
+}
+
+export type GetAuthMethodsResponse = {
+  success: true,
+  methods: AuthMethods,
+} | {
+  success: false,
+  error: string,
+}
+
+export class SNMfaService extends PureService {
+  constructor(
+    private challengeService: ChallengeService,
+  ) {
+    super()
+  }
+
+  public async handleMfa(
+    authMethodsResponse: HttpResponse<GetAuthMethodsResponse>,
+  ): Promise<MfaResult> {
+
+    if (authMethodsResponse.data?.success) {
+      const {totp} = authMethodsResponse.data.methods
+
+      if (totp) {
+        const {mfaSecretUuid} = totp
+
+        const enteredMfaCode = await this.promptForMfaValue()
+
+        if (enteredMfaCode === undefined) {
+          return {
+            status: MfaStatus.ErrorMissingCode,
+          }
+        }
+        
```

Comment:
What's the guideline on whitespace? When should one put whitespace and not put whitespace between lines?

Reply:
Here I'm conforming to @karolsojko's suggestion from some time ago to put an empty line before `return`. Makes no difference to me.

Usual guidelines are to separate logically related sections with one blank line[[1]](https://google.github.io/styleguide/javaguide.html#s4.6.1-vertical-whitespace)[[2]](https://www.python.org/dev/peps/pep-0008/#blank-lines).

This method had some blank lines leftover from refactoring (55, 61) which shall be deleted in the next update.